filebeat/input/filestream/internal/input-logfile/harvester.go

```diff
@@ -41,86 +42,117 @@ type Harvester interface {
-	ctx.Cancelation = harvesterCtx
-
-	resource, err := hg.manager.lock(ctx, s.Name())
-	if err != nil {
-		cancelHarvester()
-		return err
-	}
+	go func() {
+		defer func() {
+			if v := recover(); v != nil {
+				err := fmt.Errorf("harvester panic with: %+v\n%s", v, debug.Stack())
+				ctx.Logger.Errorf("Harvester crashed with: %+v", err)
+			}
+		}()
 
-	if _, ok := hg.readers[s.Name()]; ok {
-		cancelHarvester()
-		log.Debug("A harvester is already running for file")
-		return nil
-	}
-	hg.readers[s.Name()] = cancelHarvester
+		defer log.Debug("Stopped harvester for file")
 
-	hg.store.UpdateTTL(resource, hg.cleanTimeout)
+		harvesterCtx, cancelHarvester := context.WithCancel(ctxtool.FromCanceller(ctx.Cancelation))
+		ctx.Cancelation = harvesterCtx
+		defer cancelHarvester()
 
-	client, err := hg.pipeline.ConnectWith(beat.ClientConfig{
-		CloseRef:   ctx.Cancelation,
-		ACKHandler: newInputACKHandler(ctx.Logger),
-	})
-	if err != nil {
-		cancelHarvester()
-		return err
-	}
+		err := hg.readers.add(s.Name(), cancelHarvester)
+		if err != nil {
+			log.Errorf("error while adding new reader to the bookkeeper %v", err)
+			return
+		}
+		defer hg.readers.remove(s.Name())
 
-	cursor := makeCursor(hg.store, resource)
-	publisher := &cursorPublisher{canceler: ctx.Cancelation, client: client, cursor: &cursor}
+		resource, err := hg.locker.Lock(ctx, s.Name())
+		if err != nil {
+			log.Errorf("Error while locking resource: %v", err)
+			return
+		}
+		defer releaseResource(resource)
 
-	go func(cancel context.CancelFunc) {
+		client, err := hg.pipeline.ConnectWith(beat.ClientConfig{
+			CloseRef:   ctx.Cancelation,
+			ACKHandler: newInputACKHandler(ctx.Logger),
+		})
+		if err != nil {
+			log.Errorf("error while connecting to output with pipeline: %v", err)
+			return
+		}
 		defer client.Close()
-		defer log.Debug("Stopped harvester for file")
-		defer cancel()
-		defer releaseResource(resource)
-		defer delete(hg.readers, s.Name())
 
-		defer func() {
-			if v := recover(); v != nil {
-				err := fmt.Errorf("harvester panic with: %+v\n%s", v, debug.Stack())
-				ctx.Logger.Errorf("Harvester crashed with: %+v", err)
-			}
-		}()
+		hg.locker.UpdateTTL(resource, hg.cleanTimeout)
+		cursor := makeCursor(hg.store, resource)
+		publisher := &cursorPublisher{canceler: ctx.Cancelation, client: client, cursor: &cursor}
 
-		err := hg.harvester.Run(ctx, s, cursor, publisher)
+		err = hg.harvester.Run(ctx, s, cursor, publisher)
 		if err != nil {
 			log.Errorf("Harvester stopped: %v", err)
 		}
-	}(cancelHarvester)
+	}()
+
 	return nil
 }
 
-// Cancel stops the running Harvester for a given Source.
-func (hg *defaultHarvesterGroup) Cancel(s Source) error {
-	if cancel, ok := hg.readers[s.Name()]; ok {
-		cancel()
-		return nil
-	}
-	return fmt.Errorf("no such harvester %s", s.Name())
+// Stop stops the running Harvester for a given Source.
+func (hg *defaultHarvesterGroup) Stop(s Source) {
+	hg.readers.remove(s.Name())
```

Comment:
This function is already "blocking" as the `readerGroup.remove` locks a mutex. The mutex is relevant when we are starting multiple readers from separate goroutines in parallel from the `fileProspector`.

Thinking about it, given that the prospector event loop should not be blocked, `Stop` is not allowed to block because I am planning to use this for stopping harvesters from the prospector.